File: sdk/go/client_test.go
package agentbazaar

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func testClient(t *testing.T, handler http.HandlerFunc) (*Client, *httptest.Server) {
	t.Helper()
	srv := httptest.NewServer(handler)
	c := New(Config{BaseURL: srv.URL, APIKey: "test-key-123"})
	t.Cleanup(srv.Close)
	return c, srv
}

func writeJSON(t *testing.T, w http.ResponseWriter, status int, v any) {
	t.Helper()
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	if err := json.NewEncoder(w).Encode(v); err != nil {
		t.Fatalf("encode: %v", err)
	}
}

func TestSendsAPIKeyHeader(t *testing.T) {
	var gotKey string
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotKey = r.Header.Get("x-api-key")
		writeJSON(t, w, http.StatusOK, map[string]any{"ok": true, "offers": []any{}})
	})
	if _, err := c.ListOffers(context.Background()); err != nil {
		t.Fatalf("ListOffers: %v", err)
	}
	if gotKey != "test-key-123" {
		t.Errorf("x-api-key = %q, want %q", gotKey, "test-key-123")
	}
}

func TestRegisterParsesAPIKeyAndSetsIt(t *testing.T) {
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/agents/register" {
			t.Errorf("path = %q", r.URL.Path)
		}
		if r.Method != http.MethodPost {
			t.Errorf("method = %q", r.Method)
		}
		writeJSON(t, w, http.StatusOK, map[string]any{
			"ok": true, "agentId": "agt_123", "apiKey": "fresh-key", "cardUrl": "https://agentbazaar.app/agents/agt_123",
		})
	})
	resp, err := c.Register(context.Background(), RegisterAgentInput{
		Name: "MyBot", WalletAccountID: "0.0.1234", Capabilities: []string{"buyer"},
	})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}
	if resp.AgentID != "agt_123" || resp.APIKey != "fresh-key" {
		t.Errorf("resp = %+v", resp)
	}
	if c.apiKey != "fresh-key" {
		t.Errorf("client apiKey = %q, want fresh-key (auto-set)", c.apiKey)
	}
}

func TestSearchBuildsQueryParams(t *testing.T) {
	var qs string
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		qs = r.URL.RawQuery
		writeJSON(t, w, http.StatusOK, map[string]any{"ok": true, "results": []any{}})
	})
	max := 5.0
	_, err := c.SearchOffers(context.Background(), SearchParams{
		Query: "translate", Capability: "text.translate", MaxPrice: &max, Asset: "USDC",
	})
	if err != nil {
		t.Fatalf("SearchOffers: %v", err)
	}
	for _, want := range []string{"q=translate", "capability=text.translate", "maxPrice=5", "asset=USDC"} {
		if !strings.Contains(qs, want) {
			t.Errorf("query %q missing %q", qs, want)
		}
	}
}

func TestListOffersParsesOffers(t *testing.T) {
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, http.StatusOK, map[string]any{
			"ok": true,
			"offers": []any{
				map[string]any{
					"id": "off_1", "agentId": "agt_1", "capability": "echo.demo",
					"title": "Echo", "priceAmount": 1.5, "priceAsset": "HBAR",
					"fulfillmentType": "inline", "escrow": true,
				},
			},
		})
	})
	resp, err := c.ListOffers(context.Background())
	if err != nil {
		t.Fatalf("ListOffers: %v", err)
	}
	if len(resp.Offers) != 1 {
		t.Fatalf("offers len = %d", len(resp.Offers))
	}
	o := resp.Offers[0]
	if o.ID != "off_1" || o.PriceAmount != 1.5 || o.PriceAsset != "HBAR" || !o.Escrow {
		t.Errorf("offer = %+v", o)
	}
}

func TestBuyReturnsPaymentRequired(t *testing.T) {
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, http.StatusPaymentRequired, map[string]any{
			"ok": true, "orderId": "o-1",
			"order":   map[string]any{"id": "o-1", "status": "pending_payment"},
			"payment": map[string]any{"amount": 2.0, "asset": "HBAR", "payTo": "0.0.999", "memo": "ab-xyz"},
		})
	})
	_, err := c.Buy(context.Background(), "off_1", map[string]any{"text": "hi"}, BuyOptions{})
	if err == nil {
		t.Fatal("expected error")
	}
	var perr *ErrPaymentRequired
	if !errors.As(err, &perr) {
		t.Fatalf("error type = %T, want *ErrPaymentRequired", err)
	}
	if perr.Payment.Amount != 2.0 || perr.Payment.PayTo != "0.0.999" || perr.Payment.Memo != "ab-xyz" {
		t.Errorf("payment = %+v", perr.Payment)
	}
	if !IsPaymentRequired(err) {
		t.Error("IsPaymentRequired = false")
	}
}

func TestAPIErrorStatus(t *testing.T) {
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, http.StatusInternalServerError, map[string]any{"ok": false, "error": "boom"})
	})
	_, err := c.ListOffers(context.Background())
	if err == nil {
		t.Fatal("expected error")
	}
	var aerr *APIError
	if !errors.As(err, &aerr) {
		t.Fatalf("error type = %T, want *APIError", err)
	}
	if aerr.StatusCode != http.StatusInternalServerError || aerr.Message != "boom" {
		t.Errorf("api error = %+v", aerr)
	}
}

func TestNotificationsLimit(t *testing.T) {
	var path string
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		path = r.URL.String()
		writeJSON(t, w, http.StatusOK, map[string]any{
			"ok": true, "agentId": "agt_1", "unread": 1,
			"notifications": []any{
				map[string]any{"id": "n_1", "agentId": "agt_1", "event": "order.created",
					"title": "New order", "message": "You have a new order", "read": false},
			},
		})
	})
	resp, err := c.ListNotifications(context.Background(), 10)
	if err != nil {
		t.Fatalf("ListNotifications: %v", err)
	}
	if path != "/api/v1/notifications?limit=10" {
		t.Errorf("path = %q", path)
	}
	if len(resp.Notifications) != 1 || resp.Notifications[0].Event != "order.created" || resp.Unread != 1 {
		t.Errorf("notifications = %+v", resp)
	}
}

func TestHealthParses(t *testing.T) {
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, http.StatusOK, map[string]any{
			"ok": true, "status": "ok", "version": "1.4.5",
			"agents": 7, "offers": 12, "orders": 34, "escrows": 5,
		})
	})
	h, err := c.Health(context.Background())
	if err != nil {
		t.Fatalf("Health: %v", err)
	}
	if h.Version != "1.4.5" || h.Agents != 7 || h.Offers != 12 {
		t.Errorf("health = %+v", h)
	}
}

func TestClientRespectsCustomTimeout(t *testing.T) {
	c := New(Config{BaseURL: "http://127.0.0.1:1", Timeout: 1, APIKey: "k"})
	if c.hc.Timeout != 1 {
		t.Errorf("timeout = %v", c.hc.Timeout)
	}
	if c.baseURL != "http://127.0.0.1:1" {
		t.Errorf("baseURL = %q", c.baseURL)
	}
}

File: sdk/rust/tests/client_test.rs
//! Integration tests for the AgentBazaar Rust SDK.
//!
//! Uses a tiny std-only mock HTTP server (no external dev-dependencies) so
//! tests run offline and fast — mirroring the Go SDK's `httptest` suite.

use agentbazaar::{
    is_payment_required, BuyOptions, Client, Config, Error, RegisterAgentInput, SearchParams,
};
use std::io::{BufRead, BufReader, Read, Write};
use std::net::{TcpListener, TcpStream};
use std::sync::mpsc;
use std::thread;

/// Spawns a single-request mock HTTP server. Returns (base_url, rx) where rx
/// receives `(request_line, headers, body)` for the captured request.
fn mock_server(
    status: u16,
    body: &'static str,
) -> (String, mpsc::Receiver<(String, String, String)>) {
    let listener = TcpListener::bind("127.0.0.1:0").unwrap();
    let addr = listener.local_addr().unwrap();
    let (tx, rx) = mpsc::channel();
    thread::spawn(move || {
        for stream in listener.incoming() {
            if let Ok(mut stream) = stream {
                handle_one(&mut stream, status, body, &tx);
            }
            break; // single request per mock
        }
    });
    (format!("http://{addr}"), rx)
}

fn handle_one(
    stream: &mut TcpStream,
    status: u16,
    body: &'static str,
    tx: &mpsc::Sender<(String, String, String)>,
) {
    let mut reader = BufReader::new(stream.try_clone().unwrap());
    let mut request_line = String::new();
    let _ = reader.read_line(&mut request_line);
    let mut headers = String::new();
    let mut content_length = 0usize;
    loop {
        let mut line = String::new();
        if reader.read_line(&mut line).unwrap_or(0) == 0 || line == "\r\n" || line == "\n" {
            break;
        }
        let lower = line.to_ascii_lowercase();
        if let Some(v) = lower.strip_prefix("content-length:") {
            content_length = v.trim().parse().unwrap_or(0);
        }
        headers.push_str(&line);
    }
    let mut req_body = vec![0u8; content_length];
    if content_length > 0 {
        let _ = reader.read_exact(&mut req_body);
    }
    let _ = tx.send((
        request_line.trim().to_string(),
        headers,
        String::from_utf8_lossy(&req_body).to_string(),
    ));
    let resp = format!(
        "HTTP/1.1 {status} X\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}",
        body.len(),
        body
    );
    let _ = stream.write_all(resp.as_bytes());
}

fn test_client(api_key: &str, status: u16, body: &'static str) -> (Client, mpsc::Receiver<(String, String, String)>) {
    let (base, rx) = mock_server(status, body);
    let c = Client::new(Config {
        base_url: Some(base),
        api_key: Some(api_key.to_string()),
        ..Default::default()
    });
    (c, rx)
}

#[test]
fn sends_api_key_header() {
    let (c, rx) = test_client("test-key-123", 200, r#"{"ok":true,"offers":[]}"#);
    let resp = c.list_offers().unwrap();
    assert!(resp.ok);
    let (_, headers, _) = rx.recv().unwrap();
    assert!(
        headers.to_ascii_lowercase().contains("x-api-key: test-key-123"),
        "headers = {headers}"
    );
}

#[test]
fn register_parses_api_key_and_sets_it() {
    let (c, rx) = test_client(
        "",
        200,
        r#"{"ok":true,"agentId":"agt_123","apiKey":"fresh-key","cardUrl":"https://agentbazaar.app/agents/agt_123"}"#,
    );
    let resp = c
        .register(RegisterAgentInput {
            name: "MyBot".into(),
            wallet_account_id: "0.0.1234".into(),
            capabilities: vec!["buyer".into()],
            ..Default::default()
        })
        .unwrap();
    assert_eq!(resp.agent_id, "agt_123");
    assert_eq!(resp.api_key, "fresh-key");
    assert_eq!(c.api_key(), "fresh-key", "client key should be auto-set");
    let (request_line, _, _) = rx.recv().unwrap();
    assert!(request_line.starts_with("POST /api/v1/agents/register"));
}

#[test]
fn search_builds_query_params() {
    let (c, rx) = test_client("", 200, r#"{"ok":true,"results":[]}"#);
    let _ = c
        .search_offers(SearchParams {
            query: Some("translate".into()),
            capability: Some("text.translate".into()),
            max_price: Some(5.0),
            asset: Some("USDC".into()),
        })
        .unwrap();
    let (request_line, _, _) = rx.recv().unwrap();
    for want in [
        "q=translate",
        "capability=text.translate",
        "maxPrice=5",
        "asset=USDC",
    ] {
        assert!(
            request_line.contains(want),
            "query {request_line} missing {want}"
        );
    }
}

#[test]
fn list_offers_parses_offers() {
    let (c, _) = test_client(
        "",
        200,
        r#"{"ok":true,"offers":[{"id":"off_1","agentId":"agt_1","capability":"echo.demo","title":"Echo","priceAmount":1.5,"priceAsset":"HBAR","fulfillmentType":"inline","escrow":true}]}"#,
    );
    let resp = c.list_offers().unwrap();
    assert_eq!(resp.offers.len(), 1);
    let o = &resp.offers[0];
    assert_eq!(o.id, "off_1");
    assert_eq!(o.price_amount, 1.5);
    assert_eq!(o.price_asset, "HBAR");
    assert!(o.escrow);
}

#[test]
fn buy_returns_payment_required() {
    let (c, _) = test_client(
        "",
        402,
        r#"{"ok":true,"orderId":"o-1","order":{"id":"o-1","status":"pending_payment"},"payment":{"amount":2.0,"asset":"HBAR","payTo":"0.0.999","memo":"ab-xyz"}}"#,
    );
    let err = c
        .buy("off_1", serde_json::json!({"text": "hi"}), BuyOptions::default())
        .unwrap_err();
    match &err {
        Error::PaymentRequired(perr) => {
            assert_eq!(perr.payment.amount, 2.0);
            assert_eq!(perr.payment.pay_to, "0.0.999");
            assert_eq!(perr.payment.memo, "ab-xyz");
        }
        other => panic!("error type = {other:?}, want PaymentRequired"),
    }
    assert!(is_payment_required(&err));
}

#[test]
fn api_error_status() {
    let (c, _) = test_client("", 500, r#"{"ok":false,"error":"boom"}"#);
    let err = c.list_offers().unwrap_err();
    match &err {
        Error::Api(a) => {
            assert_eq!(a.status_code, 500);
            assert_eq!(a.message, "boom");
        }
        other => panic!("error type = {other:?}, want Api"),
    }
}

#[test]
fn delete_offer_unit() {
    let (c, rx) = test_client("k", 200, r#"{"ok":true}"#);
    c.delete_offer("off_9").unwrap();
    let (request_line, _, _) = rx.recv().unwrap();
    assert!(request_line.starts_with("DELETE /api/v1/offers/off_9"));
}
